middleware/2fa.js

```diff
@@ -0,0 +1,287 @@
+                kind: method.kind,
+                detail: method.detail,
+                requestId: -1
+            });
+        }
+    }
+    // check if 2fa method matches existing recovery method
+    const [recoveryMethod] = await account.getRecoveryMethods({ where: {
+        kind: kind.split('2fa-')[1],
+        detail,
+    }});
+    if (!recoveryMethod) {
+        // client waits to deploy contract until code is verified
+        const securityCode = await sendCode(method, twoFactorMethod);
+        const body = {
+            success: true,
+            message: '2fa initialized and code sent to verify method',
+        };
+        if (testing) {
+            body.securityCode = securityCode;
+        }
+        ctx.body = body;
+    } else {
+        // client should deploy contract
+        ctx.body = {
+            success: true, confirmed: true, message: '2fa initialized and set up using recovery method verification'
+        };
+    }
+};
+// http post http://localhost:3000/2fa/send accountId=mattlock method:='{"kind":"2fa-email","detail":"matt@near.org"}'
+// http post https://near-contract-helper-2fa.onrender.com/2fa/send accountId=mattlock method:='{"kind":"2fa-email","detail":"matt@near.org"}'
+// Call anytime after calling initCode to resend a new code, the new code will overwrite the old code
+const sendNewCode = async (ctx) => {
+    const { accountId, method, requestId, data } = ctx.request.body;
+    const [account] = await models.Account.findOrCreate({ where: { accountId } });
+    if (!account) {
+        console.warn('account should be created');
+        ctx.throw(401);
+    }
+    let [twoFactorMethod] = await account.getRecoveryMethods({ where: {
+        kind: {
+            [Op.startsWith]: '2fa-'
+        },
+    }});
+    if (!twoFactorMethod) {
+        console.warn('2fa not enabled');
+        ctx.throw(401);
+    }
+
+    await sendCode(method, twoFactorMethod, requestId, data);
+    ctx.body = {
+        success: true, message: '2fa code sent'
+    };
+};
+// http post http://localhost:3000/2fa/verify accountId=mattlock securityCode=430888
+// http post https://near-contract-helper-2fa.onrender.com/2fa/verify accountId=mattlock securityCode=437543
+// call when you want to verify the "current" securityCode
+const verifyCode = async (ctx) => {
+    const { accountId, securityCode, requestId, testing } = ctx.request.body;
+    const account = await models.Account.findOne({ where: { accountId } });
+    const [twoFactorMethod] = await account.getRecoveryMethods({ where: {
+        securityCode,
+        kind: {
+            [Op.startsWith]: '2fa-'
+        },
+        // only verify codes that are 5 minutes old (if testing make this impossible)
+        updatedAt: {
+            [Op.gt]: Date.now() - (!testing ? 300000 : -1000)
+        }
+        // cannot test for requestId equality with negative integer???
+    }});
+    // checking requestId here with weak equality (no type match)
```

Comment:
let's not have  comments which literally describe the same thing as next line of code. `twoFactorMethod.requestId != requestId` already says same thing. More interesting generally is why is code doing smth.